src/data/finsupp/witnesses.lean

```diff
@@ -0,0 +1,182 @@
+
+@[simp] lemma diff_eq_empty : f.diff g = ∅ ↔ f = g :=
+begin
+  refine ⟨λ h, _, λ h, h ▸ by simp [diff]⟩,
+  ext a,
+  refine not_ne_iff.mp (λ fg, not_ne_iff.mpr h _),
+  refine finset.ne_empty_of_mem (_ : a ∈ _),
+  simp only [diff, finset.mem_filter, finset.mem_union, finsupp.mem_support_iff],
+  exact ⟨fg.ne_or_ne _, fg⟩,
+end
+
+@[simp] lemma nonempty_diff_iff : (f.diff g).nonempty ↔ f ≠ g :=
+finset.nonempty_iff_ne_empty.trans diff_eq_empty.not
+
+@[simp]
+lemma diff_zero_right : f.diff 0 = f.support :=
+by simp only [diff, coe_zero, pi.zero_apply, support_zero, finset.union_empty,
+    finset.filter_true_of_mem, mem_support_iff, imp_self, implies_true_iff]
+
+@[simp]
+lemma diff_zero_left : (0 : α →₀ N).diff f = f.support :=
+diff_comm.trans diff_zero_right
+
+lemma subset_map_range_diff {M} [decidable_eq M] [has_zero M] {F : N → M} (F0 : F 0 = 0) :
+  (f.map_range F F0).diff (g.map_range F F0) ⊆ f.diff g :=
+begin
+  refine λ x, _,
+  simp only [mem_diff, map_range_apply, not_not],
+  exact not_imp_not.mpr (λ h, h ▸ rfl),
+end
+
+lemma map_range_diff_eq {M} [decidable_eq M] [has_zero M]
+  {F : N → M} (F0 : F 0 = 0) (hF : function.injective F) :
+  (f.map_range F F0).diff (g.map_range F F0) = f.diff g :=
+by { ext, simpa only [mem_diff] using hF.ne_iff }
+
+end N_has_zero
+
+lemma add_diff_add_eq_left [add_left_cancel_monoid N] (f : α →₀ N) {g h : α →₀ N} :
+  (f + g).diff (f + h) = g.diff h  :=
+begin
+  ext,
+  simp only [diff, ne.def, add_right_inj, finset.mem_filter, finset.mem_union, mem_support_iff,
+    coe_add, pi.add_apply, and.congr_left_iff],
+  exact λ bc, ⟨λ h, ne.ne_or_ne 0 bc, λ h, ne.ne_or_ne _ ((add_right_inj _).not.mpr bc)⟩,
+end
+
+--  can this proof by replaced by the previous one applied to `Nᵃᵒᵖ` and interlaced with `op unop`?
+lemma add_diff_add_eq_right [add_right_cancel_monoid N] {f g : α →₀ N} (h : α →₀ N):
+  (f + h).diff (g + h) = f.diff g  :=
+begin
+  ext,
+  simp only [diff, ne.def, add_left_inj, finset.mem_filter, finset.mem_union, mem_support_iff,
+    coe_add, pi.add_apply, and.congr_left_iff],
+  exact λ bc, ⟨λ h, ne.ne_or_ne 0 bc, λ h, ne.ne_or_ne _ ((add_left_inj _).not.mpr bc)⟩,
+end
+
+@[simp] lemma diff_neg_left [add_group N] {f g : α →₀ N} :
+  (- f).diff g = f.diff (- g) :=
+begin
+  nth_rewrite 0 ← neg_neg g,
+  exact map_range_diff_eq neg_zero neg_injective,
+end
+
+@[simp] lemma diff_eq_support_sub [add_group N] {f g : α →₀ N} :
+  f.diff g = (f - g).support :=
+by rw [← add_diff_add_eq_right (- g), add_right_neg, diff_zero_right, sub_eq_add_neg]
+
+end diff
+
+section wit
+variables [nonempty α] [linear_order α]
+
+section N_has_zero
+variables [has_zero N] {f g : α →₀N}
+
+/--  Given two finitely supported functions `f g : α →₀ N`, `finsupp.wit f g` is an element of `α`.
+It is a "generic" element of `α` (namely, `nonempty.some _ : α`) if and only if `f = g`.
+Otherwise, it is `a` if `a : α` is the smallest value for which `f a ≠ g a`. -/
+noncomputable def wit (f g : α →₀ N) : α :=
```

Comment:
We do care about the `is_empty α` case though, don't we? Finitely supported functions are still a linear order in that case. I think it might be best to redefine this so that it takes in a proof of `f ≠ g` as an argument.

Reply:
That said, I'm not so sure about having this as a standalone definition. It seems pretty useless outside of the proof you mention, unless I'm missing something.

Reply:
Note that `f ≠ g` implies `nonempty α`: when I made this definition, I was explicitly trying to get the typeclass mechanism to provide the proof of inequality, rather than burdening the user with it.

You can take a look at [this proof](https://github.com/leanprover-community/mathlib/blob/b428aab9f1d9b05ba0d4fc73a4e15326b7204812/src/data/finsupp/lex.lean#L153) to see how to use `wip` and dealing with `nonemptiness`.

Reply:
I split off the definition and basic lemmas for `diff` in a separate PR.

I agree that `wit` is more specialized, but avoiding giving it a name seems to me somewhat inconvenient for the application that I have in mind.  I could move the definition in the file where it is used, if that seems better.